msu/systems/serialization/metadata_emulator.nut

```diff
@@ -40,56 +50,186 @@
+		::MSU.requireOneFromTypes(["float", "integer"], _value);
 		this.__setValue(_key, _value);
 	}
 
-	function getFloat( _key )
+	function getString( _key )
 	{
-		return this.__getValue(_key);
+		if (_key in this.Data || this.RealMeta == null)
+			return this.__getValue(_key);
+		return this.RealMeta.getString(_key);
 	}
 
-	function setFloat( _key, _value )
+	function setString( _key, _value )
 	{
-		::MSU.requireFloat(_value);
+		::MSU.requireString(_value);
 		this.__setValue(_key, _value);
 	}
 
-	function hasData( _key )
+	function __hasDataRaw( _key )
 	{
 		return _key in this.Data;
 	}
 
-	function getName() // dummy, doesn't return useful value
-	{
-		return "";
-	}
-
-	function getModificationDate() // dummy, doesn't return useful value
+	function hasData( _key )
 	{
-		return "";
+		// not sure if RealMeta hasData should be checked here
+		// this could cause problems if something sets a key in metadata
+		// and doesn't set it if it already exists in RealMeta
+		// because we would then not know that we need to serialize that key
+		// the problem I have right now is with types
+		// we don't know the type of the value in RealMeta
+		// so we don't know which get<XYZ> to use
+		return this.__hasDataRaw(_key) || this.RealMeta.hasData(_key);
 	}
 
-	function getCreationDate() // dummy, doesn't return useful value
+	function getName()
 	{
-		return "";
+		if (this.RealMeta == null || this.RealMeta.getName() == null)
+			return this.Name;
+		return this.RealMeta.getName();
 	}
 
-	function getFileName() // dummy, doesn't return useful value
+	function getVersion()
 	{
-		return "";
+		if (this.RealMeta == null || this.RealMeta.getVersion() == null)
+			return this.Version;
+		return this.RealMeta.getVersion();
+	}
+
+	function getModificationDate()
+	{
+		if (this.RealMeta == null || this.RealMeta.getModificationDate() == null)
+			return this.ModificationDate;
+		return this.RealMeta.getModificationDate();
+	}
+
+	function getCreationDate()
+	{
+		if (this.RealMeta == null || this.RealMeta.getCreationDate() == null)
+			return this.CreationDate;
+		return this.RealMeta.getCreationDate();
+	}
+
+	function getFileName()
+	{
+		if (this.RealMeta == null || this.RealMeta.getFileName() == null)
+			return this.FileName;
+		return this.RealMeta.getFileName();
+	}
+
+	function __setUpDummyAssetsAndReturnCleanup()
+	{
+		local cleanup;
+		if ("Assets" in ::World && !::MSU.isNull(::World.Assets))
+		{
+			cleanup = function() {};
+		}
+		else
+		{
+			local assets = ::new("scripts/states/world/asset_manager")
+			::World.Assets <- ::WeakTableRef(assets);
+			cleanup = function() {
+				::World.Assets = null;
+				::Stash = null;
+				assets = null;
+			};
+		}
+		return cleanup;
+	}
+
+	function serialize( _out )
+	{
+		if (::MSU.System.Serialization.MidOnBeforeSerialize)
+		{
+			throw "Can't serialize metadata during onBeforeSerialize!";
+		}
+		local realMeta = this.RealMeta;
+		local knownValues = {};
+		local dummyMeta = {
```

Comment:
@Enduriel I remember we added some comments here explaining the reasoning for overwriting functions. The comments are present in the commit on your other branch, but apparently wasn't copied over to the commit on this branch when you cherrypicked/pushed.

Reply:
Yeah, probably worth adding them back in